cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -0,0 +1,194 @@
+
+    HttpResponse response = restClient.execute(HttpMethod.POST, clientConfig.resolveURL("runtime/monitor/topics"),
+                                               GSON.toJson(topicsToMonitor), Collections.emptyMap(), null);
+
+    if (response.getResponseCode() == HttpURLConnection.HTTP_UNAVAILABLE) {
+      throw new ServiceUnavailableException(response.getResponseBodyAsString());
+    }
+
+    Map<String, String> configToTopic = GSON.fromJson(response.getResponseBodyAsString(StandardCharsets.UTF_8),
+                                                      MAP_STRING_STRING_TYPE);
+
+    // TODO initialize from offset table for a given runtimeId
+    for (Map.Entry<String, String> entry : configToTopic.entrySet()) {
+      topicsToRequest.put(entry.getKey(), new MonitorConsumeRequest(entry.getValue(), entry.getKey(), null, limit));
+    }
+  }
+
+  @Override
+  protected void run() {
+    runThread = Thread.currentThread();
+
+    try {
+      while (!stopped) {
+        try {
+          if (topicsToRequest.isEmpty()) {
+            initializeTopics();
+          }
+
+          HttpResponse response = restClient
+            .execute(HttpRequest.builder(HttpMethod.POST, clientConfig.resolveURL("runtime/metadata"))
+                       .withBody(GSON.toJson(topicsToRequest)).build());
+
+          if (response.getResponseCode() == HttpURLConnection.HTTP_UNAVAILABLE) {
+            throw new ServiceUnavailableException(response.getResponseBodyAsString());
+          }
+
+          Map<String, List<MonitorMessage>> monitorResponses =
+            GSON.fromJson(response.getResponseBodyAsString(StandardCharsets.UTF_8), MAP_STRING_MESSAGE_TYPE);
+
+          processResponse(monitorResponses);
+        } catch (Exception e) {
+          OUTAGE_LOG.warn("Failed to fetch monitoring data from runtime {}, run {}. Will be retried in next iteration.",
+                          runtimeId.getProgram(), runtimeId.getRun(), e);
+        }
+        Thread.sleep(pollTimeMillis);
+      }
+    } catch (InterruptedException e) {
+      // Interruption means stopping the service.
+    }
+  }
+
+  private void processResponse(Map<String, List<MonitorMessage>> monitorResponses) throws Exception {
+    for (Map.Entry<String, MonitorConsumeRequest> request : topicsToRequest.entrySet()) {
+      publish(request.getKey(), request.getValue(), monitorResponses.get(request.getValue().getTopic()));
+    }
+  }
+
+  private void publish(String topicConfig, MonitorConsumeRequest request,
+                       List<MonitorMessage> messages) throws Exception {
+    if (messages.isEmpty()) {
+      return;
+    }
+
+    // TODO publish messages transactionally along with offset table updates
+    messagePublisher.publish(NamespaceId.SYSTEM.getNamespace(), cConf.get(topicConfig),
+                             Iterators.transform(messages.iterator(), new Function<MonitorMessage, byte[]>() {
```

Comment:
Use the `publish` method that takes `Iterator<String>` instead. You can provide the charset to the method.

Reply:
BTW, in Java 8, better use stream than guava for  iterator transformation:

```java
messages.stream().map(s -> s.getBytes(StandardCharsets.UTF_8)).iterator()
```